source/kernel/src/services/sdmmc.rs

```diff
@@ -0,0 +1,523 @@
+            .cmd(Command {
+                index: 3,
+                argument: 0,
+                options: HardwareOptions::None,
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Short,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { value, .. } => Ok(RelativeCardAddress(value)),
+            Response::Long(_) => Err(Error::from(ErrorKind::Response)),
+        }
+    }
+
+    /// Toggle the card between stand-by and transfer state
+    pub async fn select(&mut self, rca: RelativeCardAddress) -> Result<CardStatus, Error> {
+        match self
+            .cmd(Command {
+                index: 7,
+                argument: rca.0,
+                options: HardwareOptions::None,
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Short,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { value, .. } => Ok(CardStatus(value)),
+            Response::Long(_) => Err(Error::from(ErrorKind::Response)),
+        }
+    }
+
+    /// Use 4 data lanes
+    pub async fn set_wide_bus(&mut self, rca: RelativeCardAddress) -> Result<CardStatus, Error> {
+        // Go to *APP* mode before sending application command
+        let _ = self
+            .cmd(Command {
+                index: 55,
+                argument: rca.0,
+                options: HardwareOptions::None,
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Short,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?;
+
+        match self
+            .cmd(Command {
+                index: 6,
+                argument: 0b10, // instruct card to use 4-bits bus
+                options: HardwareOptions::SetBusWidth(BusWidth::Quad),
+                kind: CommandKind::Control,
+                rsp_type: ResponseType::Short,
+                rsp_crc: true,
+                buffer: None,
+            })
+            .await?
+        {
+            Response::Short { value, .. } => Ok(CardStatus(value)),
+            Response::Long(_) => Err(Error::from(ErrorKind::Response)),
+        }
+    }
+
+    /// Read the desired number of data blocks into the provided buffer, starting at the given sector.
+    pub async fn read(
+        &mut self,
+        sector: u32,
+        blocks: usize,
+        buf: FixedVec<u8>,
+    ) -> Result<FixedVec<u8>, Error> {
+        // The provider buffer should have space for the requested amount of data
+        if buf.capacity() < 512 * blocks {
+            return Err(Error::from(ErrorKind::Buffer));
+        }
+
+        match self
+            .cmd(Command {
+                index: 18,
+                argument: sector,
+                options: HardwareOptions::None,
+                kind: CommandKind::Read(512 * blocks),
```

Comment:
nit, not a blocker: maybe:
```suggestion
        const BLOCK_SIZE: usize = 512;
        
        let bytes = BLOCK_SIZE * blocks;
        // The provider buffer should have space for the requested amount of data
        if buf.capacity() < bytes {
            return Err(Error::from(ErrorKind::Buffer));
        }

        match self
            .cmd(Command {
                index: 18,
                argument: sector,
                options: HardwareOptions::None,
                kind: CommandKind::Read(bytes),
```